[kernel 2c/n] Introduce `kernel::Context`, encapsulate global init/teardown

dongcarl opened the pull request:
The full `init/common.cpp` is dependent on things like ArgsManager (which we wish to remove from libbitcoinkernel in the future) and sanity checks. These aren't necessary for libbitcoinkernel so we only extract the portion that is necessary (namely `init::{Set,Unset}Globals()`.

dongcarl committed:
test: Use Set/UnsetGlobals in BasicTestingSetup

...instead of calling initialization functions directly and having to
keep around a ECCVerifyHandle member variable.

This makes the initialization codepath of our tests more closely
resemble those of AppInitMain and potentially eases the review of
subsequent commit removing init::{Set,Unset}Globals.

[META] In a future commit, we will introduce a kernel::Context which
       calls init::{Set,Unset}Globals in its ctor and dtor. It will be
       owned by node::NodeContext, so in the end, this patchset won't
       have made the previously local ECCVerifyHandle global.

MarcoFalke commented on src/Makefile.am:
```diff
@@ -870,7 +872,7 @@ libbitcoinkernel_la_SOURCES = \
   index/blockfilterindex.cpp \
   index/coinstatsindex.cpp \
-  init/common.cpp \
+  kernel/init/common.cpp \
```
Not sure if we want a `kernel` folder for code that is shared between node and kernel. An alternative would be to name it `init/kernel.cpp` or `init/globals.cpp`?

Otherwise it will be difficult to decide whether to put kernel code that is used elsewhere as well in `kernel` or in the "normal" hierarchy?  cc @ryanofsky

dongcarl replied:
I think the main thing here is to have a name that clearly indicates that this contains "only the common initialization code that the kernel needs" so that people don't accidentally add non-kernel code to it (it would result in a linker error in the libbitcoinkernel CI job anyway).

Other PRs in the project (e.g. #24410) does this as well, since "extracting only what the kernel needs" is the main goal of this phase of the project. I like using the `kernel/` prefix since it makes the aforementioned indication clear, and doesn't require inventing a new name.

In this particular case though, perhaps `kernel/init.cpp` would work.

One question is: do we keep the two functions moved in the `init::` namespace or the `kernel::` namespace.

In the longer run, I think eventually (very far off when `libbitcoinkernel` is very minimal) we could put everything that links into libbitcoinkernel into the `src/kernel/` folder.

MarcoFalke replied:
So is the goal to eventually split validation into kernel/validation.cpp and node/validation.cpp ? What about util functions that are included in the kernel?

dongcarl replied:
> So is the goal to eventually split validation into kernel/validation.cpp and node/validation.cpp ?

Yes, far off into the future we will likely do that (although I think in reality `node/validation.cpp` will likely be either: 1. a very very short file, or 2. deleted because what would be left belongs better in a differently named file).

> What about util functions that are included in the kernel?

I've discussed this with Ryan in #24352, and we came to the conclusion that we will move all the utils that `libbitcoinkernel` depends on to `libbitcoin_util`, now that's just w/re the libraries, we could do either one of two things for the hierarchy:

1. `src/kernel/` and `src/util/`, or
2. `src/kernel/` and `src/kernel/util/`

Either is fine with me. See here for more discussion: https://github.com/bitcoin/bitcoin/pull/24352#issuecomment-1048231861

dongcarl replied:
> so I am wondering if there should be two util libs? If yes, then it would make sense to have both `src/util` and `src/kernel/util`.  If not, then I don't care what is done.

I think that the point of https://github.com/bitcoin/bitcoin/pull/24352#issuecomment-1048231861 is to establish exactly what you're saying. In effect:

- `common` will be the non-kernel "util lib" (i.e. `src/util`), and
- `util` will be the kernel "util lib" (i.e. `src/kernel/util`).

laanwj replied:
> I think that the point of https://github.com/bitcoin/bitcoin/pull/24352#issuecomment-1048231861 is to establish exactly what you're saying. In effect:

SGTM

> For example logging is used

Yes. Though, for the eventual libbitcoin kernel we need to find some way to do logging that doesn't make the client application depend on bitcoind's logging subsystem (say, a hook / notification registration). Out of scope here, of course.

MarcoFalke commented on src/init/common.h:
```diff
@@ -8,11 +8,11 @@
 #ifndef BITCOIN_INIT_COMMON_H
 #define BITCOIN_INIT_COMMON_H
 
+#include <kernel/include/init/common.h>
```
nit: Pretty sure this is illegal according to iwyu?

dongcarl replied:
Ah right. Yeah the thought here (in general for the `libbitcoinkernel` project) is for the existing headers to export things from the "kernel-extracted version of headers" so that we don't have too much code churn adding `#include <kernel/*.h>` all over the place. (See my other comment here for more details: https://github.com/bitcoin/bitcoin/pull/25065#discussion_r865910970)

I think a good solution would be to do:
```
#include <kernel/include/init/common.h> // IWYU pragma: export
```

dongcarl replied:
Fixed as of: 31c21f3644f6f42f6e4c3c231e95bf0b87ab86f5

dongcarl commented:
Pushed d5876d1efbf01f45fff2b42e0bcf2bc12eb25240 -> 13e65511dba702d115b7bea147b12eefaea9d53f
- Adjusted according to https://github.com/bitcoin/bitcoin/pull/24410#issuecomment-1118984576

dongcarl commented:
Pushed 13e65511dba702d115b7bea147b12eefaea9d53f -> 708e2711e591ce4958900715348cb0313763f2cb
- Addressed: https://github.com/bitcoin/bitcoin/pull/25065#discussion_r865578361

dongcarl commented:
Pushed 708e2711e591ce4958900715348cb0313763f2cb -> e4c35528e58a74138aa7e0560dd7f59f1086c765
- Fixed incorrect header ifdef: https://github.com/bitcoin/bitcoin/runs/6313105188

dongcarl commented:
Pushed e4c35528e5 -> e6f8606007
- Rebased over master after merge of #25064

dongcarl commented:
Push e6f8606007a1690dcf9265d15c7fe15cf5eb5239 -> 6d93c443a44f06296e8014c65bc70f44cc494fca
- Rebased over master after merge of #24410

theuni reviewed (commented):
Concept ACK and looks good from a light review.

No comment on the naming/structure, what was discussed above SGTM. I suspect this will change around a few times before finally settling anyway.

The split here seems pretty arbitrary... I could imagine a few different approaches to chopping up common.cpp and (without digging too much) this isn't the most obvious one to me. But if this keeps things moving, sure 👍.

ryanofsky commented:
Few suggestions (feel free to ignore):

- Would squash 3 commits to 1 commit. There's like 7 lines of code moving in this PR. I don't think moving it in 3 parts instead of 1 part helps anyone!
- Would rename `src/kernel/init/common.h,cpp` to `src/kernel/init.h,cpp`
- Would rename `namespace init` to `namespace kernel` (namespace name not matching directory name is like a sign pointing in wrong direction, probably worse than not having the sign)
- Maybe for followup, instead of `kernel::SetGlobals` `kernel::UnsetGlobals` call these `kernel::Init` `kernel::Destroy` or `Load`/`Unload` or similar. If you wanted to be really fancy you could call these `kernel::Context::Context()` and `kernel::Context::~Context()` but no need to go crazy with another context struct yet.

ryanofsky commented:
To explain suggestions in https://github.com/bitcoin/bitcoin/pull/25065#issuecomment-1136567868 a little, I think the libbitcoin_kernel external interface should not use global variables. libbitcoin_kernel is supposed to be a library, and having used many libraries, I'd claim that libraries which use global variables or other implicit singletons are a pain to develop and test with.

I understand bitcoin code uses global variables, and you don't want to make big changes to existing code to get rid of them. This is fine. But we shouldn't add kernel functions which require the implementation use to global variables, and make them difficult to get rid of in the future.

You can easily create a clean API that does not require other code to use global variables by defining a `kernel::Context` struct that owns whatever state is needed to run kernel functions, and passing `Context*` to functions in `src/kernel/` meant to be called externally. Internally (especially at first), `Context*` won't be accessible to a lot of existing code, so global variables should still be used. They just need to point to the context or to specific members of the context instead of owning the state directly. Not much existing code should have to change. The only code that might have to change is the lowest-level code directly referencing global variables that were moved. Only some of this code would have to change, and no code above it would have to change.

At a higher level, the `kernel::Context` struct could just be understood as a thing which lets libbitcoin_kernel code remember state between calls, and interact with the outside world without relying on globals. So it can ask where is the data dir? How do I log this message? How do I generate a random number? What time is it? And so on

dongcarl committed:
init: Initialize globals with kernel::Context's life

...instead of explicitly calling init::{Set,Unset}Globals.

Cool thing about this is that in both the testing and bitcoin-chainstate
codepaths, we no longer need to explicitly unset globals. The
kernel::Context goes out of scope and the globals are unset
"automatically".

Also construct kernel::Context outside of AppInitSanityChecks()

dongcarl committed:
kernel: Introduce empty and unused kernel::Context

[META] In the next commit, we will move the init::{Set,Unset}Globals
       logic into this struct.

Co-Authored-By: Ryan Ofsky <ryan@ofsky.org>

dongcarl commented:
Pushed 6d93c443a44f06296e8014c65bc70f44cc494fca -> 9bc9309bd61f36ac68b41fe2328d2e3f505397e1
- Overhauled PR to introduce a `kernel::Context` and move the global initialization logic to its ctor/dtor. (Thanks Cory+Ryan)

dongcarl commented on src/kernel/init/common.cpp:
```diff
@@ -0,0 +1,30 @@
+
+#include <memory>
+
+static std::unique_ptr<ECCVerifyHandle> globalVerifyHandle;
```
Fixed as of 9bc9309bd61f36ac68b41fe2328d2e3f505397e1

dongcarl commented on src/init/common.h:
```diff
@@ -8,11 +8,11 @@
 #ifndef BITCOIN_INIT_COMMON_H
 #define BITCOIN_INIT_COMMON_H
 
+#include <kernel/init/common.h> // IWYU pragma: export
```
No longer relevant as of 9bc9309bd61f36ac68b41fe2328d2e3f505397e1

dongcarl commented:
Thanks for all the feedback @ryanofsky and @theuni, I really appreciate them. I totally agree that in this case it was a great idea and easy to introduce the `kernel::Context` along with setting the globals, I actually discovered I had an old branch that did exactly this 😅 

Although in this case it was quite easy to make this change so we get the N I C E C O D E, it might be the case that in future PRs there will be instances where getting the N I C E C O D E will require very intrusive changes that doesn't serve the goal of Phase 1, which is just about decoupling modules. Should these cases arise, I hope we can decide to table them for later (although most of these cases are likely solved by introducing a `kernel::Context g_kernel`, but we'll do that when we get to it).

ryanofsky reviewed (approved):
Code review ACK 9bc9309bd61f36ac68b41fe2328d2e3f505397e1. Looks great, I'd say.

> there will be instances where getting the N I C E C O D E will require very intrusive changes that doesn't serve the goal

This seems fine as long as we have a reasonable idea of what a good cleaned up state would look like. Would just want to avoid making a mess without a plan to clean it up.

ryanofsky commented on src/init/common.h:
```diff
@@ -8,11 +8,11 @@
 
 namespace init {
-void SetGlobals();
-void UnsetGlobals();
```
re: https://github.com/bitcoin/bitcoin/pull/25065#discussion_r882023913

> Doing that does mean we can't remove `compat/glibcxx_sanity.cpp` from libbitcoinkernel

Looking at current `SanityChecks()` function, maybe a bigger problem moving it into the kernel is that it is using
`InitError` from  `ui_interface.cpp`. So might be worth postponing this or doing it separately if it would complicate this PR.

dongcarl replied:
Fixed as of 9ac87854af62d271df2f383d8f2f798c70453448

theuni replied:
Nice, thanks.

Now that we have the notion of a context, it would be possible to either:
- Make it impossible to create an invalid context (sanity checks called at construction time)
- Allow for "invalid" contexts, which api calls check for.

But those can be done much later when we start actually defining an api.

dongcarl committed:
kernel: SanityChecks: Return an error struct

This reduces libbitcoinkernel's coupling with ui_interface and
translation.

dongcarl commented:
Push 9bc9309bd61f36ac68b41fe2328d2e3f505397e10 -> 9ac87854af62d271df2f383d8f2f798c70453448
- Implemented: https://github.com/bitcoin/bitcoin/pull/25065#discussion_r882101216

-----

@ryanofsky I know you don't like overly-split commits, happy to merge the last two if you wish.

theuni commented on src/kernel/checks.h:
```diff
@@ -5,14 +5,23 @@
 
+enum class SanityCheckError {
+    ERROR_INITITALIZING_ECC,
+    ERROR_GLIBCXX_INSANITY,
```
Though it's quite literal here, keeping with precedent, let's not reintroduce this language: https://github.com/bitcoin/bitcoin/pull/5727

The errors are already namespaced as part of the enum class anyway, how about just `SanityCheckError::ERROR_GLIBCXX` ?

dongcarl replied:
Fixed as of 69500384a2f7f3d63151f23daee873878cc74c0b

dongcarl commented:
Pushed 9ac87854af62d271df2f383d8f2f798c70453448 -> 69500384a2f7f3d63151f23daee873878cc74c0b
- Addressed https://github.com/bitcoin/bitcoin/pull/25065#discussion_r882189863

vasild commented on src/bitcoin-chainstate.cpp:
```diff
@@ -49,7 +51,7 @@ int main(int argc, char* argv[])
     const CChainParams& chainparams = Params();
 
-    init::SetGlobals(); // ECC_Start, etc.
+    kernel::Context context{};
```
The variable of type `kernel::Context` is called `context` here and `kernel` inside `struct NodeContext`. `kernel_context`, `kernel_state` or `kernel_ctx` may be better.

dongcarl replied:
Fixed in 319031b9798f1ed8f0a06e679284be26e0b42c92

vasild replied:
Now it is called `kernel_context` here and `kernel` elsewhere. Not a big deal, but I think it is better to be consistent if there are no other reasons for the naming.

vasild commented on src/kernel/checks.cpp:
```diff
@@ -0,0 +1,34 @@
+
+namespace kernel {
+
+std::optional<SanityCheckError> SanityChecks(Context&)
```
Why bother introducing an argument if it is going to be unused?

ryanofsky replied:
> Why bother introducing an argument if it is going to be unused?

It's because src/kernel/ is supposed to provide a library API, and the API should not have to change as we remove globals internally.

dongcarl replied:
Going to mark this as resolved, feel free to reopen if not!

vasild commented on src/kernel/checks.h:
```diff
@@ -0,0 +1,28 @@
+
+struct Context;
+
+enum class SanityCheckError {
+    ERROR_ECC,
```
The word "error" is redundant in the enum name and in each enum value: 

```
SanityCheckError::ERROR_ECC
           ^^^^^  ^^^^^
```

Maybe strip the `ERROR_` prefix: `SanityCheckError::ECC`?

vasild reviewed (approved):
ACK 239501f5d0e1df8d36a62f10c4de4e9cdc913a9c

Some non-blockers below.

theuni commented on src/init.cpp:
```diff
@@ -1091,11 +1091,11 @@ static bool LockDataDirectory(bool probeOnly)
     // ********************************************************* Step 4: sanity checks
 
-    init::SetGlobals();
+    node.kernel = std::make_unique<kernel::Context>();
```
Mmm, this whole function is a mess and doesn't really make much sense anymore.

I really don't like the creation of the `kernel::Context` happening in here. That's not at all intuitive. I would much prefer to see the caller create and set it. Mind addressing that here?

A few potential follow-ups for a later PR:
- I'm hoping that the SanityChecks move into the `kernel::Context` creation as a next step.
- I think it may make sense for `NodeContext` to take a `kernel::Context` in its constructor so that it always exists in the NodeContext. That might be much easier said than done though, I haven't looked deeply into it.
- Those changes would leave only `LockDataDirectory` here, meaning that `AppInitSanityChecks` would just go away.

dongcarl replied:
> I would much prefer to see the caller create and set it. Mind addressing that here?

Let me know what you think about: 8d63a934cf4053851ba79d4a44326cd079bc8288

dongcarl replied:
> A few potential follow-ups for a later PR:
> ...

I will open an issue after this one's merged so we don't lose track of these!

theuni replied:
Ah yes, this is much more straightforward imo. +1 for squashing that in.

dongcarl commented on src/kernel/context.h:
```diff
@@ -0,0 +1,31 @@
+class ECCVerifyHandle;
+
+namespace kernel {
+//! Context struct holding global the kernel library's global state, and passed
```
Fixed in cae459cd9aa0da220da92e11b3a645b9980cdaf5

dongcarl commented on src/kernel/context.h:
```diff
@@ -0,0 +1,31 @@
+//!
+//! State stored directly in this struct should be simple. More complex state
+//! should be stored to std::unique_ptr members pointing to opaque types.
+struct Context {
```
Fixed in 4fa8d0b11aefa06f548013a35bbcfa9fdbed81ff

dongcarl commented:
Pushed cf4c2043f592afe1d175f4c54e9728da38c0b184 -> 8d63a934cf4053851ba79d4a44326cd079bc8288
- Addressed various review comments

dongcarl committed:
Move init::SanityCheck to kernel::SanityCheck

dongcarl commented:
Pushed 8d63a934cf4053851ba79d4a44326cd079bc8288 -> ded2ef302cd29d3dc88a847f0fe8c947b3a6044a
- Squashed in 8d63a934cf4053851ba79d4a44326cd079bc8288

theuni commented on src/bitcoin-chainstate.cpp:
```diff
@@ -49,7 +51,7 @@ int main(int argc, char* argv[])
     const CChainParams& chainparams = Params();
 
-    init::SetGlobals(); // ECC_Start, etc.
+    kernel::Context kernel_context{};
```
This also needs a `kernel::SanityChecks(kernel_context)`, no? Another argument for making it part of context creation.

(It was also missing before, but presumably because it was still tangled up with the ui).

theuni replied:
Btw, I'm loving how this just untangled itself. Great suggestion from @ryanofsky [here](https://github.com/bitcoin/bitcoin/pull/25065#discussion_r882101216)!

dongcarl replied:
Fixed as of: fa2ee7b12c0c15be8e9807444443b5bd2c91c3e5

theuni replied:
Completely agree it's wonky, I had a similar complaint here: https://github.com/bitcoin/bitcoin/pull/25065#discussion_r883689055

I think it's fine as-is for now though, as context creation will definitely be changing in subsequent PRs.

vasild reviewed (approved):
ACK ded2ef302cd29d3dc88a847f0fe8c947b3a6044a

I would expect the kernel API to be defined in `src/kernel/*.h` inside `namespace kernel` and those files to not define anything else. `src/kernel/chainstatemanager_opts.h` does not fit that pattern - it defines `ChainstateManagerOpts` outside of the `kernel` namespace. So, is `ChainstateManagerOpts` part of the kernel API? If yes, then it should be inside `namespace kernel`, if not then it should not be defined in `src/kernel/*.h`. `ChainstateManagerOpts` is only used in `validation.h`.

dongcarl commented:
@vasild Your comment inspired me to make a disclaimer, I think I'm going to use it in every `libbitcoinkernel` PR from now on!

![kernel-disclaimer](https://user-images.githubusercontent.com/3445290/171222306-274faa75-87ca-4a85-a938-ff05bc129940.png)

dongcarl commented:
Pushed ded2ef302cd29d3dc88a847f0fe8c947b3a6044a -> fa2ee7b12c0c15be8e9807444443b5bd2c91c3e5
- Rebased over merge of #25233
- Addressed https://github.com/bitcoin/bitcoin/pull/25065#discussion_r884001856
- Fixed some `#include`s in `src/bitcoin-chainstate.cpp`

dongcarl commented:
> I would expect the kernel API to be defined in `src/kernel/*.h` inside `namespace kernel` and those files to not define anything else. `src/kernel/chainstatemanager_opts.h` does not fit that pattern - it defines `ChainstateManagerOpts` outside of the `kernel` namespace. So, is `ChainstateManagerOpts` part of the kernel API? If yes, then it should be inside `namespace kernel`, if not then it should not be defined in `src/kernel/*.h`. `ChainstateManagerOpts` is only used in `validation.h`.

Sorry about my cheeky post above 😄  To respond to your comment: yes it should eventually be in `namespace kernel`, so should most of `validation.cpp`. Haven't done it yet because it's not required by any of the "PRs in the main thrust of libbitcoinkernel".

theuni commented on src/node/interfaces.cpp:
```diff
@@ -90,8 +90,14 @@ class NodeImpl : public Node
     {
-        return AppInitBasicSetup(gArgs) && AppInitParameterInteraction(gArgs, /*use_syscall_sandbox=*/false) && AppInitSanityChecks() &&
-               AppInitLockDataDirectory() && AppInitInterfaces(*m_context);
+        return AppInitBasicSetup(gArgs) &&
```
Heh, this has gotten unwieldy and hard to read imo.

How about (untested)
```c++
if (!AppInitBasicSetup(gArgs)) return false;
if (!AppInitParameterInteraction(gArgs, /*use_syscall_sandbox=*/false)) return false;

m_context->kernel = std::make_unique<kernel::Context>();
if (!AppInitSanityChecks(*m_context->kernel)) return false;
if (!AppInitLockDataDirectory()) return false;
if (!AppInitInterfaces(*m_context)) return false;
return true;
```

theuni reviewed (commented):
I took one last pass through, looking good.

Left a comment about a messy function, not the end of the world if it doesn't get addressed here.

I do think the const ref issues (there may be others) should be addressed though so that it's clear moving forward what's intended to be read-only.

vasild reviewed (approved):
ACK fa2ee7b12c0c15be8e9807444443b5bd2c91c3e5

> yes it should eventually be in `namespace kernel`

Ok, why not get it right from the first time, since it is trivial to do so:

```diff
--- i/src/kernel/chainstatemanager_opts.h
+++ w/src/kernel/chainstatemanager_opts.h
@@ -7,17 +7,20 @@
 
 #include <cstdint>
 #include <functional>
 
 class CChainParams;
 
+namespace kernel {
+
 /**
  * An options struct for `ChainstateManager`, more ergonomically referred to as
  * `ChainstateManager::Options` due to the using-declaration in
  * `ChainstateManager`.
  */
 struct ChainstateManagerOpts {
     const CChainParams& chainparams;
     const std::function<int64_t()> adjusted_time_callback{nullptr};
 };
+} // namespace kernel
 
 #endif // BITCOIN_KERNEL_CHAINSTATEMANAGER_OPTS_H
--- i/src/validation.h
+++ w/src/validation.h
@@ -854,13 +854,13 @@ private:
         const CBlockHeader& block,
         BlockValidationState& state,
         CBlockIndex** ppindex) EXCLUSIVE_LOCKS_REQUIRED(cs_main);
     friend CChainState;
 
 public:
-    using Options = ChainstateManagerOpts;
+    using Options = kernel::ChainstateManagerOpts;
 
     explicit ChainstateManager(const Options& opts)
         : m_chainparams{opts.chainparams},
           m_adjusted_time_callback{Assert(opts.adjusted_time_callback)} {};
 
     const CChainParams& GetParams() const { return m_chainparams; }
```

dongcarl commented:
Pushed fa2ee7b12c0c15be8e9807444443b5bd2c91c3e5 -> a8cdeca079a362d2a021aaf7387fbb75a15ebc18
- Addressed
  - https://github.com/bitcoin/bitcoin/pull/25065#discussion_r886809654
  - https://github.com/bitcoin/bitcoin/pull/25065#discussion_r886810741
  - https://github.com/bitcoin/bitcoin/pull/25065#discussion_r886815297

dongcarl commented:
> Ok, why not get it right from the first time, since it is trivial to do so:

Yeah sorry I should have done that in the first place. But it's unrelated to the current PR right now so we'll just do it later. I've added it as a TODO to the [umbrella issue](https://github.com/bitcoin/bitcoin/issues/24303)!

dongcarl commented on src/init.cpp:
```diff
@@ -1089,13 +1091,23 @@ static bool LockDataDirectory(bool probeOnly)
+            break;
+        case kernel::SanityCheckError::ERROR_CHRONO:
+            InitError(Untranslated("Clock epoch mismatch. Aborting."));
+            break;
```
Addressed in d87784ac87364fc977bbf9769c8bdb72dea8cbf9, feel free to reopen this convo if not resolved!

theuni replied:
FWIW, gcc/clang have built-ins (`__builtin_unreachable`) for [expressing this explicitly](https://gcc.gnu.org/onlinedocs/gcc/Other-Builtins.html).  

LLVM even has a [macro for it](https://github.com/llvm/llvm-project/blob/main/llvm/include/llvm/Support/ErrorHandling.h#L143), and specifies the use of `llvm_unreachable` in its [coding standard](https://llvm.org/docs/CodingStandards.html).

Maybe we should borrow the idea and add a similar macro. Not in this PR though, of course :)

MarcoFalke replied:
Interesting, though I think we'd always want to abort, even in release builds. For example, it would be UB to skip `return` in a non-void function, no? This is also what the docs say:

> llvm_unreachable() becomes an optimizer hint that the current location is not supposed to be reachable: the hint turns such code path into undefined behavior.

https://github.com/llvm/llvm-project/blob/main/llvm/include/llvm/Support/ErrorHandling.h#L131

For reference, in https://github.com/bitcoin/bitcoin/pull/24812#discussion_r851597471 I argued that `assert(0)` is good enough for us. Also, I don't think there is a compiler that doesn't understand `assert(0)`?

MarcoFalke replied:
I think the primary question here was why there is no default case, to which the answer is that we happily accept a fallthrough to an InitError. No objection to a macro that just calls `assert(0)`, if ppl think it is useful.

vasild replied:
> ... why there is no default case ...

My favorite reason is that, because the switch value is an enum, we get a compile-time warning (or error with `-Werror` which is used by CI) if not all enum values are listed.

theuni commented on src/test/util/setup_common.h:
```diff
@@ -81,7 +81,7 @@ static constexpr CAmount CENT{1000000};
  */
 struct BasicTestingSetup {
-    node::NodeContext m_node;
+    node::NodeContext m_node; // keep as first member to be destructed last
```
A foot-gun-proof way to do this would be to inherit from a class that holds a node::NodeContext.

dongcarl replied:
Something like this?

```diff
diff --git a/src/node/context.h b/src/node/context.h
index 31be308787..7ed1292180 100644
--- a/src/node/context.h
+++ b/src/node/context.h
@@ -29,6 +29,16 @@ class Init;
 class WalletLoader;
 } // namespace interfaces
 
+namespace {
+/**
+ * Dummy NodeContext to ensure that kernel::Context is always destructed last.
+ */
+struct BaseNodeContext {
+    //! libbitcoin_kernel context
+    std::unique_ptr<kernel::Context> kernel;
+};
+}
+
 namespace node {
 //! NodeContext struct containing references to chain state and connection
 //! state.
@@ -40,9 +50,7 @@ namespace node {
 //! and make code more modular and testable. The struct isn't intended to have
 //! any member functions. It should just be a collection of references that can
 //! be used without pulling in unwanted dependencies or functionality.
-struct NodeContext {
-    //! libbitcoin_kernel context
-    std::unique_ptr<kernel::Context> kernel;
+struct NodeContext : BaseNodeContext {
     //! Init interface for initializing current process and connecting to other processes.
     interfaces::Init* init{nullptr};
     std::unique_ptr<AddrMan> addrman;
diff --git a/src/test/util/setup_common.h b/src/test/util/setup_common.h
index 5c31cfc22b..87d5d4a24b 100644
--- a/src/test/util/setup_common.h
+++ b/src/test/util/setup_common.h
@@ -77,12 +77,20 @@ static inline bool InsecureRandBool() { return g_insecure_rand_ctx.randbool(); }
 
 static constexpr CAmount CENT{1000000};
 
+namespace {
+/**
+ * Dummy TestingSetup to ensure that node::NodeContext (and by extension
+ * kernel::Context) is always destructed last.
+ */
+struct BaseTestingSetup {
+    node::NodeContext m_node;
+};
+}
+
 /** Basic testing setup.
  * This just configures logging, data dir and chain parameters.
  */
-struct BasicTestingSetup {
-    node::NodeContext m_node; // keep as first member to be destructed last
-
+struct BasicTestingSetup : BaseTestingSetup {
     explicit BasicTestingSetup(const std::string& chainName = CBaseChainParams::MAIN, const std::vector<const char*>& extra_args = {});
     ~BasicTestingSetup();
 
```

theuni replied:
Yep, pretty much!

Sorry, I didn't mean to imply that needed to be done as part of this PR though. It might be worth taking a look at doing it generically with a template, which could help communicate the intent more clearly:
```c++
BasicTestingSetup : MustDestructLast<node::NodeContext>
```

But maybe that's over-engineering, idk.

ryanofsky replied:
> A foot-gun-proof way to do this would be to inherit from a class that holds a node::NodeContext.

Would really encourage not doing this. There is no ambiguity about initialization order in the c++ language: earlier members are constructed first and destructed last, later members are constructed later and destructed earlier. Moving members into a base class to control what they are initialized here would seems mostly like a form of obfuscation.

It is true that if a future developer willingly wants to ignore a "this member should be first" comment it will be harder for them to do that if inheritance is involved, but it's true in general that if you obfuscate code it is harder to modify, and I don't think this logic leads to good places

theuni replied:
@ryanofsky Points taken. I think the intents/relationships need to be made clear somehow. Programmatically is one way, but maybe a comment works just as well without all the obfuscation.

In any case, this definitely shouldn't slow down this PR.

ryanofsky replied:
I do agree relationship should be explicit if one thing depends on the other. Another safeguard that may be useful is if thing A needs to be destroyed after thing B, thing B destructor can assert that thing A is not destroyed yet (assuming relevant code will already have test coverage since it is part of the test setup)

dongcarl replied:
Okay, sounds like we're okay with leaving it as just a comment like in the latest push d87784ac87364fc977bbf9769c8bdb72dea8cbf9. Going to close this convo but feel free to re-open if further discussion is worth it.

theuni reviewed (approved):
ACK d87784ac87364fc977bbf9769c8bdb72dea8cbf9

There are a few good outstanding suggestions, but this work is such a continuum that I think it's fine to address them in subsequent PRs if that's easier for @dongcarl.

vasild reviewed (approved):
ACK d87784ac87364fc977bbf9769c8bdb72dea8cbf9

MarcoFalke commented on src/bitcoin-chainstate.cpp:
```diff
@@ -49,7 +52,11 @@ int main(int argc, char* argv[])
 
-    init::SetGlobals(); // ECC_Start, etc.
+    kernel::Context kernel_context{};
+    // We can't use a goto here, but we can use an assert since none of the
```
unrelated: Probably a good time to get rid of `goto` instead of adding comments on why `goto` can't be used here. :sweat_smile:

maflcko reviewed (commented):
Left some nits and a test setup question.

MarcoFalke commented on src/kernel/context.h:
```diff
@@ -0,0 +1,31 @@
+namespace kernel {
+//! Context struct holding the kernel library's logically global state, and
+//! passed to external libbitcoin_kernel functions which need access to this
+//! state. The kernel libary API is a work in progress, so state organization
```
```
src/kernel/context.h:15: libary ==> library

fanquake replied:
being done in #25307

fanquake commented on src/kernel/checks.cpp:
```diff
@@ -0,0 +1,33 @@
+// Copyright (c) 2022 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
```
All of `src/kernel` is now under IWYU.

fanquake commented on src/kernel/context.cpp:
```diff
@@ -0,0 +1,33 @@
+    LogPrintf("Using the '%s' SHA256 implementation\n", sha256_algo);
+    RandomInit();
+    ECC_Start();
+    ecc_verify_handle.reset(new ECCVerifyHandle());
```
This code no-longer exists. Was removed in the most recent secp256k1 update (#26691).
